### ISSUE TITLE
Regexes with some lookaround assertions match an extra character 

### ISSUE THREAD

**Issue by codesections** [rakudo/rakudo#4512]
Many regexes that start with [lookaround assertions](https://docs.raku.org/language/regexes#Lookaround_assertions) that match both a character in the input string and an "empty string"-ish character class match one more character than they should:

```raku
'aa' ~~ /<?[\s a]> . /    # OUTPUT: ｢aa｣ 
'ab' ~~ /<?[\N a]>/       # OUTPUT: ｢a｣
```

I am not sure if this issue would be more appropriate for the NQP repo; if so, let me know and I can transfer it.  See also [a StackOverflow question](https://stackoverflow.com/questions/69004383/lookaround-regex-and-character-consumption) discussing the issue.

**Comment by ugexe** [comment]
Lookaround char classes like `<?[\s a]>` get compiled into a tree structure that wraps the individual parts in a container node, but the regex compiler ignores the zero width flag on that container so each part still consumes a character. A fix would be to  propagate the zero-width flag down to the individual parts. Similarly for negative lookarounds like `<![\s a]>` the negation flag is also ignored on the container, so a fix would be to rewrite the structure so negation is applied to each part individually.

[Here](https://github.com/rakudo/rakudo/compare/main...ugexe:rakudo:ugexe/issue-4512) is a prototype of the aforementioned changes that give the following results:
```
$ raku -e 'say "aa" ~~ /<?[\s a]> ./;'
｢a｣

$ raku -e 'say "bb" ~~ /<![\s a]> ./;'
｢b｣
```

**Comment by jubilatious1** [comment]
Just to understand....the zero-width lookaround only properly worked with a single character, prior to this fix?

(previous discussions centered on backslashed characters within bespoke character classes as being the culprit).

**Comment by ugexe** [comment]
No, it was about pieces not characters. Something like `<?[abc]>` was fine, but any class built from two or more parts (`[\s a]`, `[a..c x..z]`, etc) lost its zero-width flag. Or from the commit message of the fix https://github.com/Raku/nqp/pull/869:

> Previously the MoarVM regex compiler only acted on the zerowidth and negate flags of leaf nodes. A lookaround over a character class that compiles to an alt or concat, such as <?[\s a]> or <![a] - [b]>, had the flags set on that group node, so its parts still consumed a character and a negated one was not inverted at all.
